controllers/capabilities/tenant_controller.go

```diff
@@ -98,6 +102,39 @@ func (r *TenantReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, err
 	}
 
+	// Tenant has been marked for deletion
+	if tenantR.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
+		// your deletion handling
+		err := controllerhelper.DeleteTenant(tenantR, portaClient)
```

Comment:
This is the most critical part of the code. It should be idempotent and expect errors at any operation.

For the initial implementation, I would say that when this op `DeleteTenant` does not return error, it means that the 3scale API confirmed that eventually the tenant will be deleted (or has been deleted). 

What is the idea to check confirmation?

Let's say this is what happens:
- Customer deletes the CR
- 3scale operator sends HTTP request to delete the tenant
- When checking to confirm, 3scale API says it is not deleted. 
 
The 3scale operator will not retry and the CR will be "tagged" to delete but actually never deleted because the finalizer was not removed

My logic proposal is:
- Customer deletes the CR
- 3scale operator sends HTTP request to delete the tenant
- If the response is DONE (200) or DOES NOT EXIST(404) or whatever http status code is 3scale API returning when trying to delete a tenant already in "scheduled for deletetion", then remove the finalizer
- Any other issue, connection issue or http 5xx response, then retry.

Reply:
Following the scenario above, I've spotted where the issue could be. 
Think this would work better than checking the error codes?
```
		// delete tenant
		err := controllerhelper.DeleteTenant(tenantR, portaClient)
		if err != nil {
			return ctrl.Result{}, err
		}

		// confirm tenant deletion
		isTenantDeleted, err := controllerhelper.ConfirmTenantDeleted(tenantR, portaClient)
		if err != nil {
			return ctrl.Result{}, err
		}

		if !isTenantDeleted {
			return ctrl.Result{Requeue: true}, nil
		}
		
		// add or remove finalizer
		err = controllerhelper.ReconcileFinalizers(tenantR, r.Client, tenantFinalizer)
		if err != nil {
			return ctrl.Result{}, err
		}

		return ctrl.Result{}, nil
		```